:bookmark: Version 0.0.1

ThierryO opened the pull request:
Initial version of the package. Currently only provides support for `flandersqmd-book`. Run the code below to get an idea of how the package works.

```
pak::pkg_install("inbo/flandersqmd@draft")
path <- tempfile()
dir.create(path)
shortname <- "test"
# to create a report skeleton
create_report(path = path, shortname = shortname, version = "gha")
# to run when you have the report numbers
insert_missing_metadata(file.path(path, shortname))
# to run when you have the cover page
report_path <- file.path(path, shortname)
```

ThierryO committed:
✨ Bump package version

ThierryO committed:
✨ Add author2yaml()

ThierryO committed:
✨ Add add_address()

ThierryO committed:
🚨 Fix checklist issues

ThierryO committed:
✨ Add add_recommendations()

ThierryO committed:
✨ Add add_index()

ThierryO committed:
✨ Add add_abstract()

ThierryO committed:
✨ Add add_intro()

ThierryO committed:
✨ Add add_bibliography()

ThierryO committed:
✨ Update the left navbar

ThierryO committed:
✨ Add create_report()

ThierryO committed:
🎨 Rename add_intro() to add_chapter()

ThierryO committed:
✨ Add insert_missing_metadata()

ThierryO committed:
🚨 Fix cyclomatic complexity

ThierryO committed:
🚨 Fix checklist issues

ThierryO committed:
✨ New add_cover() function

ThierryO committed:
📝 Update README

ThierryO committed:
✨ New post_render() function

ThierryO committed:
🐛 Ensure the affiliation stays a list

ThierryO committed:
🐛 Replace GNU url with SPDX url

ThierryO committed:
💚 Use devel version of pak in GHA

ThierryO committed:
♻️ Create an internal store_yaml() function

ThierryO committed:
✨ Create new function add_author()

ThierryO committed:
🐛 Correctly update the affiliations in fix_affiliation()

ThierryO committed:
🐛 Update post_render() so it works with quarto preview

janneadolf commented on R/create_report.R:
```diff
@@ -0,0 +1,238 @@
+#'  Defaults to the current working directory.
+#' @param shortname The name of the report project.
+#' The location of the folder `shortname` depends on the content of `path`.
+#' When `path` is a subfolder of a git repository, it is changed to the root
```
Print message that informs users about this change?

janneadolf commented on README.md:
```suggestion
Though optional, we strongly recommend to create a `checklist`
```

ThierryO replied:
Good suggestion. But I need to update it in `README.Rmd`.

janneadolf replied:
Ok, right.

janneadolf commented on README.md:
Useful to add the information that a quarto book project is created as a result of using flandersqmd::create_report() and also link to the respective documentation?

ThierryO replied:
Good suggestion. Where would you add that information? As a comment in the `md` files and/or the `_quarto.yml`?

janneadolf replied:
I was more thinking about adding a paragraph to the README, will make a suggestion to clarify.

ThierryO replied:
Updated in b288fea8a68a3ff0e0c31778b015a00510357bf7

janneadolf reviewed (commented):
I successfully executed the following steps (see [test repo](https://github.com/janneadolf/test-flandersqmd)):

- install flandersqmd (under R 4.5.0 with INBO setup; flanderqmd would not build without rtools updates under R 4.4.3)
- create a new report
- update metadata (leaving entries blank)
- render report via quarto::quarto_render

I left a few suggestions/comments attached to individual files.

**A few (additional) specific suggestions concerning flandersqmd::create_report**

- During report creation the user can select among 2 corporate identity options: entity, Flanders; the first option is not very informative, sensible to change to my_entity?
- During report creation the user has to provide a "short title used for the file name"; sensible to specify that this concerns the file name of the pdf output file?
- Maybe print a message to inform users about the change of path, especially as  the user is given the impression of control by having to specify it (see help file: "When path is a subfolder of a git repository, it is changed to the root of that git repository")?
- Section files in the report skeleton are of format .md but could (in some cases) better be .qmd per default?

**General personal impression**
I prefer to install the flandersqmd quarto extensions directly in combination with the manual setup (see [readme](https://github.com/inbo/flandersqmd-book/blob/main/README.md)). This route is more transparent to me with respect to setup (e.g., in which folder is the quarto project created, which files are extension files / report source files / output files / files that are "by-products" of rendering and can be removed to keep the folder clean). I also like to have more (immediate) control over the setup.

ThierryO commented:
Thanks for the feedback.

> * install flandersqmd (under R 4.5.0 with INBO setup; flanderqmd would not build without rtools updates under R 4.4.3)

Not building under R 4.4.3 is most likely due to the fact that you don't have the compatible Rtools installed.

> * During report creation the user can select among 2 corporate identity options: entity, Flanders; the first option is not very informative, sensible to change to my_entity?

In the future there will be more entities than only INBO. Then the function will first ask which entity, then which level. I could ask which entity, but that seems a bit silly given that currently there is only one option.

> * During report creation the user has to provide a "short title used for the file name"; sensible to specify that this concerns the file name of the pdf output file?

It will also affect the URL on the INBO website (in a later stage).

> * Maybe print a message to inform users about the change of path, especially as  the user is given the impression of control by having to specify it (see help file: "When path is a subfolder of a git repository, it is changed to the root of that git repository")?

I will simplify that. When `path` is a `checklist` project, it will use the checklist default of `source/shortname` (project) or `inst/shortname` (package). Otherwise it will use `path`.

> * Section files in the report skeleton are of format .md but could (in some cases) better be .qmd per default?

`qmd` is only needed when using code in the file. I would expect that those files seldom will contain code. Or am I missing something?

> **General personal impression** I prefer to install the flandersqmd quarto extensions directly in combination with the manual setup (see [readme](https://github.com/inbo/flandersqmd-book/blob/main/README.md)). This route is more transparent to me with respect to setup (e.g., in which folder is the quarto project created, which files are extension files / report source files / output files / files that are "by-products" of rendering and can be removed to keep the folder clean). I also like to have more (immediate) control over the setup.

You always can choose either to do it manually or via `create_report`. Note that you can use the other functions even when you created the report manually.

janneadolf commented:
> > * During report creation the user has to provide a "short title used for the file name"; sensible to specify that this concerns the file name of the pdf output file?
> 
> It will also affect the URL on the INBO website (in a later stage).

Probably useful to inform users about that as well then?

> > * Maybe print a message to inform users about the change of path, especially as  the user is given the impression of control by having to specify it (see help file: "When path is a subfolder of a git repository, it is changed to the root of that git repository")?
> 
> I will simplify that. When `path` is a `checklist` project, it will use the checklist default of `source/shortname` (project) or `inst/shortname` (package). Otherwise it will use `path`.

Sounds good.
 
> > * Section files in the report skeleton are of format .md but could (in some cases) better be .qmd per default?
> 
> `qmd` is only needed when using code in the file. I would expect that those files seldom will contain code. Or am I missing something?

That is true. On the other hand, users might expect at least a single .qmd file as this package supports the use of quarto extensions. Also, using .qmd is not (overly) costly even if there are no code chunks or snippets involved, or is it?
Alternative could be to include an example code chunk in one of the section files, or include a section file called 'reportbody' or similiar with an example code chunk?  

> > **General personal impression** I prefer to install the flandersqmd quarto extensions directly in combination with the manual setup (see [readme](https://github.com/inbo/flandersqmd-book/blob/main/README.md)). This route is more transparent to me with respect to setup (e.g., in which folder is the quarto project created, which files are extension files / report source files / output files / files that are "by-products" of rendering and can be removed to keep the folder clean). I also like to have more (immediate) control over the setup.
> 
> You always can choose either to do it manually or via `create_report`. Note that you can use the other functions even when you created the report manually.

That is an interesting variation. Would it be worthwile to mention this option in the readme file and/or reference the extensions repo(s) and mention the option of manual installation and setup?

janneadolf commented on README.Rmd:
```diff
@@ -29,40 +29,69 @@ knitr::opts_chunk$set(
 
-What is special about using `README.Rmd` instead of just `README.md`? 
-You can include R chunks like so:
+Though optional, we strongly recommend to first create a `checklist` project first.
```
```suggestion
Though optional, we strongly recommend to create a `checklist` project first.
```

janneadolf commented on README.Rmd:
```diff
@@ -29,40 +29,69 @@ knitr::opts_chunk$set(
-You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date.
-`devtools::build_readme()` is handy for this.
+Then generate a skeleton for the report using `create_report()`.
+The function will guide you interactively through the process of setting up a report.
```
```suggestion
The function will guide you interactively through the process of setting up a report.
Once the setup is completed the report will automatically be rendered in both html and pdf format.
The respective output files are located in the output folder of the report folder.
```

janneadolf commented on README.Rmd:
````diff
@@ -29,40 +29,69 @@ knitr::opts_chunk$set(
+library(flandersqmd)
+create_report(file.path(path, "flandersqmd-book"), shortname = "myreport")
+```
+
````
Potentially add something along the following lines:
"The environment you are operating in when resuming work on the generated report skeleton is a quarto book project automatically created by `create_report()`. You can find more information on quarto book projects in the [quarto documentation](https://quarto.org/docs/books/)."

janneadolf commented on README.Rmd:
````diff
@@ -29,40 +29,69 @@ knitr::opts_chunk$set(
+)
+# Render the report
+quarto::quarto_render(file.path(path, "flandersqmd-book", "myreport"))
+```
````
Add some more verbose information about rendering including option to render via R Studio Build pane within report-related quarto book project?

janneadolf reviewed (commented):
Added a few follow-up/additional comments. 

An additional thought re the functionality of `create_report()`: Would it make sense to add an argument that allows control over whether the skeleton is automatically rendered upon creation?

ThierryO committed:
Apply suggestions from code review

Co-authored-by: Janne Adolf <77746642+janneadolf@users.noreply.github.com>

ThierryO committed:
📝 Improve documentation of create_report()

ThierryO committed:
♻️ Create_report() only changes the path of checklist projects

ElsLommelen commented:
I just did some quick tests, I already give the results and I'll have another try tomorrow.

In the first attempt I accidentally made a mistake in one of the questions, and it broke down without saving anything:
![image](https://github.com/user-attachments/assets/6dcf65fa-a47f-4a36-a7f0-5e36de160fd8)
I was still in the beginning, and I didn't answer a lot of questions yet, but this could be frustrating if it would happen after a lot of questions, I think.  Is it feasible (and not too much work) to capture this kind of problems?

In the second attempt I filled in all information, and I got an error for the rendering, but on the other hand the report was present and the html rendered in the temp file.
The error:
![image](https://github.com/user-attachments/assets/777f05f7-5dc8-492a-9245-8e68f6670757)
So the error message is clear, and I got an html output, but the error message could be a bit scary for someone that is not familiar with this stuff.  (I instinctively checked the folder, but if you don't, you might think everything went wrong.)

And also a remark/question from the saved authors.  In an earlier test for protocolhelper, I added some authors with English affiliation names, but now I choose for a Dutch report.  I didn't know how to deal with it and left the English names, but unfortunately they appear in English in the qmd.  Of course I can easily change them afterwards, or update the affiliation each time, but is there an easier way to switch between Dutch and English affiliation names?

(To be continued...)

ThierryO commented:
> In the first attempt I accidentally made a mistake in one of the questions, and it broke down without saving anything: ![image](https://private-user-images.githubusercontent.com/18716476/453561507-6dcf65fa-a47f-4a36-a7f0-5e36de160fd8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk2MjY5ODEsIm5iZiI6MTc0OTYyNjY4MSwicGF0aCI6Ii8xODcxNjQ3Ni80NTM1NjE1MDctNmRjZjY1ZmEtYTQ3Zi00YTM2LWE3ZjAtNWUzNmRlMTYwZmQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjExVDA3MjQ0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3OWMxZWE4NWJlYzdmZGU1YTkzNDkzODJlMzYxYmMzZjk0ZmFmMDVkMTg0NzU3MjJjNjUwYzFiNDY1ZjViMWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.QJtm428LlQVXo4hnFgmqgnL2eh3n_BPJI4QZX2EN-aA) I was still in the beginning, and I didn't answer a lot of questions yet, but this could be frustrating if it would happen after a lot of questions, I think. Is it feasible (and not too much work) to capture this kind of problems?

This is something that we need to fix in [`checklist`](https://github.com/inbo/checklist/blob/6af592f5617c6f4e0a695d65375170c833a5c017/R/setup_project.R#L294)

> 
> In the second attempt I filled in all information, and I got an error for the rendering, but on the other hand the report was present and the html rendered in the temp file. The error: ![image](https://private-user-images.githubusercontent.com/18716476/453568124-777f05f7-5dc8-492a-9245-8e68f6670757.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk2MjY5ODEsIm5iZiI6MTc0OTYyNjY4MSwicGF0aCI6Ii8xODcxNjQ3Ni80NTM1NjgxMjQtNzc3ZjA1ZjctNWRjOC00OTJhLTkyNDUtOGU2OGY2NjcwNzU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjExVDA3MjQ0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyMjYxZWE0OWE5MTE1ZmNiMGYwMzJjOWRmMDQxNWIzMjg1ZDRhNWJkY2M0NTMwNmVlZjAxZGNiZTYyZDU1MWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lts2TkWPRKSf2QBHYHV97lKATZZWvr4M0GOmsejtpy8) So the error message is clear, and I got an html output, but the error message could be a bit scary for someone that is not familiar with this stuff. (I instinctively checked the folder, but if you don't, you might think everything went wrong.)

This might be a good argument for not rendering the document. The user would still get the error when rendering but then it is more clear that it is due to the rendering.

> 
> And also a remark/question from the saved authors. In an earlier test for protocolhelper, I added some authors with English affiliation names, but now I choose for a Dutch report. I didn't know how to deal with it and left the English names, but unfortunately they appear in English in the qmd. Of course I can easily change them afterwards, or update the affiliation each time, but is there an easier way to switch between Dutch and English affiliation names?

I'm working on that in a new version of checklist. The new version will insert the affiliation based on the language of the document, provided that the organisation name is given in that language. Otherwise it will use the default organisation name.

> 
> (To be continued...)

ElsLommelen commented:
> This might be a good argument for not rendering the document. The user would still get the error when rendering but then it is more clear that it is due to the rendering.

Actually I only get the error when rendering the pdf, the html builds without any problems.  So if a user only needs the html, there is even no need to worry about this issue.  So it may indeed be a good idea to not render the document, but then make sure the function gives a message where the project is saved.

> I'm working on that in a new version of checklist. The new version will insert the affiliation based on the language of the document, provided that the organisation name is given in that language. Otherwise it will use the default organisation name.

That would be great!

I continued testing, and I noticed it is not that easy to add metadata with `insert_missing_metadata()` when working in the project:
![image](https://github.com/user-attachments/assets/26edb904-b7b9-40ce-9b6e-88f578bce7d5)
As I realised a temp folder might not be the right place to do such test (as the link might not be accessible after closing the session), I also tested it with a test project in a 'normal' folder on my computer, which gave exacly the same errors.  As people are likely to work in their project when adding metadata afterwards, it would be nice to have an 'easy' way to add these without having to type the path manually.

In this new test I also had a look at some of the questions.  On the whole it works fine, is easy in use, but I found some minor issues:

When I choose the same name as an author and a reviewer, I get the remark `xxx is already listed as author` (see below), which is good, but it is followed by a list of names without clear instruction.  It may be useful to add `Please select another reviewer` here.
![image](https://github.com/user-attachments/assets/a3138fe9-1fb5-44cc-a17e-f57606ec1208)

In the rendered version I get dots in the institute name after adding metadata with `insert_missing_metadata()`, but to be honest, I did not render this version before adding metadata.  In another test where I did not add metadata, this problem did not appear.  The metadata in the _quarto.yml file do not have these dots.
![image](https://github.com/user-attachments/assets/9f56932b-c8bb-4e68-bae5-df51924e933c)



```
> # to run when you have the cover page
> report_path <- file.path(path, shortname)
```
I tried this advice that was in the beginning of this PR, but I have no clue how this should be done.  I first tried it in the repo and session where I ran the rest of the script, then in the repo of the quarto report, but it only adds a variable, I don't understand how it could change the document.  Of course I tried rendering the report again after running this command, but as no changes were detected, the rendering was not done again.  And also with a small change to trigger the rendering, I see no difference in the html.  (As mentioned before, the pdf is not build due to the error, and as this kind of errors often occur in Windows, a lot of users will be confronted with this.)

So on the whole it is easy in use and helps a lot to setup a quarto report, but it still has some minor issues.  (And I explored a bit, I did not test every detail, so I'm not sure I found them all.)

ThierryO committed:
🚸 Improve error message when the reviewer is one of the authors

ThierryO committed:
🚸 Use the current working directory as default report_path

ThierryO committed:
🐛 Insert_missing_metadata() correctly updates the yaml

ThierryO commented:
> I continued testing, and I noticed it is not that easy to add metadata with `insert_missing_metadata()` when working in the project:

`report_path` must be the path of the project. I've set the default to `"."` in fd123ebcd939ed926292cfb2c34b7f17f855b8c4. Now you can run `insert_missing_metadata()` when `_quarto.yml` is in the working directory.

> When I choose the same name as an author and a reviewer, I get the remark `xxx is already listed as author` (see below), which is good, but it is followed by a list of names without clear instruction. It may be useful to add `Please select another reviewer` here.

Fixed in fe1d0ea93bb17c55b7f95ca068773e9beb1da726

> In the rendered version I get dots in the institute name after adding metadata with `insert_missing_metadata()`, but to be honest, I did not render this version before adding metadata. In another test where I did not add metadata, this problem did not appear. The metadata in the _quarto.yml file do not have these dots.

Fixed in 27e66bcd6840e6925952ae69773cd54d6a016414

> 
> ```
> > # to run when you have the cover page
> > report_path <- file.path(path, shortname)
> ```
> 

That should have been `add_cover(report_path, cover = "cover.pdf")`

ThierryO committed:
✏️ Fix typo

ThierryO committed:
📝 Add draft of workshop presentation

ThierryO committed:
🔧 Ignore flandersqmd extensions leftovers

ThierryO committed:
🚨 Fix linter

ThierryO committed:
✏️ Add custom dictionary

ElsLommelen commented:
Thanks for solving these minor issues, the WD as default `report_path` in all functions makes it a lot easier to use!

For the cover: what is the difference between adding a path or url in `insert_missing_metadata()` and the separate function `add_cover()`?  Is it just a matter of adding everything together after receiving the metadata versus adding the cover at another moment?  Is the cover only added in the .pdf version, or in the .html version as well?  I get the impression that the cover file should be placed in the report folder before running the script.  It would be helpful to make this clear in the documentation, and maybe also an informative error or warning?
![image](https://github.com/user-attachments/assets/fd94a369-87ce-430d-81e1-90fb9d5783ad)

After making a report and adding metadata with `insert_missing_metadata()`, I rendered the report again.  While year and doi were added to most of the pages in the html (cover, samenvatting, recommendations and bibliography), it was not added to the Introduction page.  This page also showed the original 'Draft' watermark, while it had changed in other pages.  So it seems this page is not updated while rendering, but I don't know if this problem can be solved within flandersqmd?
![image](https://github.com/user-attachments/assets/f52c4efb-4fe9-4e1b-b5e4-66a46c8d8e09)
(I did not add any additional chapters before testing, it might be useful to test this as well after addressing this problem.)

ThierryO commented:
> For the cover: what is the difference between adding a path or url in `insert_missing_metadata()` and the separate function `add_cover()`?

This is due to the INBO workflow. You first must provide a document will all the metadata (except for the cover). And then they will create the cover for you.

> After making a report and adding metadata with `insert_missing_metadata()`, I rendered the report again. While year and doi were added to most of the pages in the html (cover, samenvatting, recommendations and bibliography), it was not added to the Introduction page.

It seems like that page wasn't rendered properly. Can you try to remove all output files and render them again.

ElsLommelen commented:
> This is due to the INBO workflow. You first must provide a document will all the metadata (except for the cover). And then they will create the cover for you.

So regarding the INBO workflow, there is no need to add this question in `insert_missing_metadata()`?



> It seems like that page wasn't rendered properly. Can you try to remove all output files and render them again.

Indeed, after removing the output files and restarting the R session, it could be rendered properly.

ThierryO commented:
> > This is due to the INBO workflow. You first must provide a document will all the metadata (except for the cover). And then they will create the cover for you.
> 
> So regarding the INBO workflow, there is no need to add this question in `insert_missing_metadata()`?

We could. But you couldn't answer all questions at the same time. And the function do other stuff internally. `insert_missing_metadata()` only has to update `_quarto.yml` whereas `add_cover()` needs to do some file conversions as well.

> 
> > It seems like that page wasn't rendered properly. Can you try to remove all output files and render them again.
> 
> Indeed, after removing the output files and restarting the R session, it could be rendered properly.

Great.

ThierryO committed:
🔥 Don't render the skeleton

ThierryO committed:
🚸 Update workshop

ThierryO committed:
🐛 Open documentation link in the workshop slides in a new tab

ThierryO committed:
📝 Mention the workshop slides in the README

ThierryO committed:
add missing partials
